Bootstrappable Builds

tobtoht opened the pull request:
This PR proposes to replace Gitian with Guix to achieve bootstrappable builds for release binaries.

### What?

https://youtube.com/watch?v=I2iShmUTEl8

If you have 15 minutes, please consider watching this presentation by Carl Dong linked above as it covers the migration (Gitian → Guix) this PR proposes to make. It explains why reproducibility alone is not enough.

Other resources:

- https://github.com/Sjors/nado-book/blob/master/attacks/guix.md
- https://bootstrappable.org

### Why?

- **Greater build system security**: Bootstrappability allows us to _audit and reproduce_ our toolchain instead of blindly _trusting_ binary downloads. Our build environment can be built from source, [all the way down](https://guix.gnu.org/en/blog/2023/the-full-source-bootstrap-building-from-source-all-the-way-down/). It allows us to reduce our supply chain attack surface by only including the packages that we need, and nothing else.

- **Easier to set up**: Guix runs on any Linux distribution and on most architectures (x86_64, aarch64, riscv64). To produce reproducible release binaries, you only need to install Guix and run the build script. This hopefully leads to more participants in the pre-release verified reproduction process.

- **More flexibility**: Unlike Gitian, we are not limited to the package set of a particular Ubuntu version. Guix allows us to pick and choose our toolchains. This allows us to use the latest compilers while targeting older versions of glibc and would, for example, make it easier to add a modern Rust compiler to our build environment. Packages that are not available in Guix can easily be defined in the manifest or upstreamed.

- **Better developer UX**: Guix allows us to modify any detail about our build environment with ease. Debugging build issues takes less time because we have shell access to the build environment. The monero source code is bind mounted into the container, so edits to package definitions can be tested incrementally.

- **Future proof**: Guix is actively developed, Gitian is in [maintenance mode](https://github.com/devrandom/gitian-builder) (serious bug fixes only).

### How to test?

Requirements:

- any Linux distribution
- 50 GB of free disk space
- 4 or more cores recommended
- 2 GB RAM per thread

```bash
# Install guix
$ apt install guix  # ubuntu 22.04, debian 11, or later 
or 
# use the install script: https://guix.gnu.org/manual/en/html_node/Binary-Installation.html

# Clone the repo
$ git clone https://github.com/tobtoht/monero
$ cd monero
$ git checkout guix

# Run the builds
$ ./contrib/guix/guix-build  # this will take several hours, uses all available cores
or
$ JOBS=N ./contrib/guix/guix-build # replace N with the desired maximum number of threads

# Obtain a list of hashes
$ uname --machine && find guix/guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
```

For more information, see README.md

### Notes

| Environment   | Guix version (NEW)         | Gitian version (OLD) |
|---------------|----------------------------|----------------------|
| Guix          | `53396a22af` (28 Aug 2024) | n/a                  |
| Glibc         | 2.27                       | 2.27                 |
| Binutils      | 2.38                       | 2.30                 |
| GCC           | 12.4.0                     | 7.5.0                |
| Clang         | 11.1.0 (darwin, freebsd)   | 9.0.0                |
| Linux headers | 6.1.106 (LTS)*              | 5.4.0                |

\* this does not affect run-time kernel requirements, see [glibc docs](https://sourceware.org/glibc/wiki/FAQ#What_version_of_the_Linux_kernel_headers_should_be_used.3F).

The `guix.yml` workflow does not use caching because it uses too much space and causes evictions for other caches (10 GB limit). We only need to run it when there is a change to `contrib/{depends,guix}`, which happens infrequently and doesn't need to complete quickly.

This PR removes `native_clang` because Guix provides us with a bootstrapped package and non-guix builds can use the system clang. This PR introduces backwards incompatible compiler flags, bumping the minimum Clang version to 10 (up from 9) for depends builds. Clang >=10 is available in the package managers of all supported distros (including Ubuntu 18.04).

The minimum glibc version for all Linux targets is now 2.27 (Ubuntu 18.04, Debian 10), for more details see https://github.com/monero-project/monero/pull/9171#issuecomment-1952926105.

The source archive now includes all submodules.

### Q&A

**What role does Guix serve in the build process?**

Guix is used to set-up a reproducible, bootstrappable, containerized build environment. This replaces the Ubuntu 18.04 based environment we used for Gitian builds.

Monero dependencies are built using the `depends` build system inside the container provided by Guix. While Guix could eventually replace `depends`, it would be too drastic for this PR. 

**Can Guix run on any Linux distribution?**

Yes, with few exceptions. Installing Guix on an immutable distro like Fedora SilverBlue might not be possible.

To install Guix, use the official install script: https://guix.gnu.org/manual/en/html_node/Binary-Installation.html

**Can Guix run on macOS or Windows?**

Not natively. You can run Guix builds on macOS inside a virtual machine (even on Apple Silicon). On Windows, it should be possible to install Guix on WSL2.

**Can Guix run on aarch64 / riscv64 machines?**

Yes. It even produces the same bit-identical release binaries as Guix run on a x86_64 machine. This isn't possible with Gitian.

Targets that don't currently build on non-x86_64: Android, FreeBSD.

**Are Guix builds slower than Gitian?**

Guix builds are only slower the first time, because it needs to build toolchains (or the entire package graph if `--no-substitutes` is used) from source. Subsequent runs just as fast (or faster) than Gitian due to caching.

**Why Guix instead of Nix?**

Nix packages are not bootstrappable.

**What does bootstrappable mean in this context?**

It means that the package graph for all packages included in our build environment is rooted in a single 357-byte heavily annotated program. We no longer have to trust binaries (downloaded from Ubuntu servers) for reproducible builds.

For more information: https://guix.gnu.org/en/blog/2023/the-full-source-bootstrap-building-from-source-all-the-way-down/

**If a new version of Guix is released, does that mean our build environment is no longer reproducible?**

No, builds remain reproducible indefinitely, assuming the source code for all packages is archived and remains available. The exact version of Guix, including all of its packages is pinned using [time-machine](https://guix.gnu.org/manual/en/html_node/Invoking-guix-time_002dmachine.html).

More information: 

- https://guix.gnu.org/blog/2024/source-code-archiving-in-guix-new-publication/
- https://guix.gnu.org/en/blog/2024/adventures-on-the-quest-for-long-term-reproducible-deployment/

**How do we define which packages are included in the build environment?**

In `manifest.scm`. In this file, we can choose packages that are already available in Guix or define packages ourselves.

**When should we define packages in the manifest?**

If a package, specific package version, or toolchain isn't available in Guix and it isn't possible to upstream or we can't we can't wait on that.

**Is the build environment identical for all targets?**

No, extra packages can be added for each target.

**If we change the time-machine commit, does that mean our build environment changes?**

Yes. The manifest generally does not specify which packages versions are included in the build environment. If a package was updated in Guix between the old and new commit, we will have the updated version in our build environment. Updating the time-machine can therefore cause breakage and should be done with caution.

**How do I get shell access to the build environment for development purposes?**

In `contrib/guix/guix-build` replace `bash -c "cd /monero && bash contrib/guix/libexec/build.sh"` with `bash`.

Then invoke with: `FORCE_DIRTY_WORKTREE=1 ./contrib/guix/guix-build`.

**Where to start review?**

`contrib/guix/README.md`
↓
`contrib/guix/guix-build`
↓
`contrib/guix/manifest.scm`
↓
`contrib/guix/libexec/build.sh`

### Troubleshooting

- Might segfault during compilation on first generation Ryzen processors due to a [hardware fault](https://old.reddit.com/r/Amd/comments/6ubmd1/ryzen_compilation_segfaults_positive_rma/).
- Ubuntu 24.04: If you run into `mount "none" on "/tmp/guix-directory.": Permission denied`  you need to create an apparmor profile for Guix, see: https://github.com/monero-project/monero/pull/8929#issuecomment-2325048697

### Follow-up PRs

- #9207
- #9440

0xFFFC0000 commented:
I cannot wait until we finalize and merge this. Thanks for your great work @tobtoht

tobtoht commented:
I don't expect to make major changes to this PR, it's ready for testing. Reviews on any of the sub-PRs would help move this forward.

tobtoht commented on .gitignore:
```diff
@@ -25,10 +25,8 @@ miniupnpcstrings.h
-contrib/gitian/docker/
-contrib/gitian/sigs/
+# guix
+/guix
```
Finished builds are available in `./guix/guix-build-<COMMIT>/`

```
├── guix-build-222ea1a30058
│   ├── build
│   │   └── distsrc-222ea1a30058-x86_64-linux-gnu  # build directory
│   ├── logs
│   │   └── x86_64-linux-gnu  # various logs for reproducibility debugging
│   │       ├── depends-gen_id-build.txt
│   │       ├── depends-gen_id-host.txt
│   │       ├── depends-hashes.txt
│   │       ├── depends-packages.txt
│   │       ├── guix-env.txt
│   │       ├── guix-hashes.txt
│   │       └── SHA256SUMS.part
│   ├── output
│   │   └── x86_64-linux-gnu
│   │       └── monero-x86_64-linux-gnu-222ea1a30058.tar.bz2  # binary tarball
│   └── var
│       ├── precious_dirs  # directories to keep when running `./contrib/guix/guix-clean`
│       └── profiles
```

tobtoht commented on contrib/depends/funcs.mk:
```diff
@@ -267,4 +275,4 @@ $(foreach package,$(all_packages),$(eval $(call int_config_attach_build_config,$
 
 #special exception: if a toolchain package exists, all non-native packages depend on it
-$(foreach package,$(packages),$(eval $($(package)_unpacked): |$($($(host_arch)_$(host_os)_native_toolchain)_cached) ))
+$(foreach package,$(packages),$(eval $($(package)_extracted): |$($($(host_arch)_$(host_os)_native_toolchain)_cached) ))
```
`_unpacked` target doesn't exist.

tobtoht commented on contrib/guix/manifest.scm:
```diff
@@ -0,0 +1,271 @@
+        gawk
+        sed
+        moreutils
+        patchelf
```
`patchelf` is unused. It is occasionally useful for debugging.

tobtoht commented on contrib/guix/libexec/build.sh:
```diff
@@ -0,0 +1,380 @@
+esac
+
+export GIT_DISCOVERY_ACROSS_FILESYSTEM=1
+export USE_DEVICE_TREZOR_MANDATORY=1
```
Force Trezor support for release binaries.

tobtoht commented on contrib/depends/funcs.mk:
```diff
@@ -1,18 +1,25 @@
-$(1)_arflags=$($($(1)_type)_ARFLAGS) $($($(1)_type)_$(release_type)_ARFLAGS)
-$(1)_ldflags=$($($(1)_type)_LDFLAGS) $($($(1)_type)_$(release_type)_LDFLAGS) -L$($($(1)_type)_prefix)/lib
-$(1)_cppflags=$($($(1)_type)_CPPFLAGS) $($($(1)_type)_$(release_type)_CPPFLAGS) -I$($($(1)_type)_prefix)/include
+$(1)_cc=$$($$($(1)_type)_CC)
+$(1)_cxx=$$($$($(1)_type)_CXX)
+$(1)_objc=$$($$($(1)_type)_OBJC)
+$(1)_objcxx=$$($$($(1)_type)_OBJCXX)
+$(1)_ar=$$($$($(1)_type)_AR)
+$(1)_ranlib=$$($$($(1)_type)_RANLIB)
+$(1)_libtool=$$($$($(1)_type)_LIBTOOL)
+$(1)_nm=$$($$($(1)_type)_NM)
+$(1)_cflags=$$($$($(1)_type)_CFLAGS) \
+            $$($$($(1)_type)_$$(release_type)_CFLAGS)
+$(1)_cxxflags=$$($$($(1)_type)_CXXFLAGS) \
+              $$($$($(1)_type)_$$(release_type)_CXXFLAGS)
+$(1)_arflags=$$($$($(1)_type)_ARFLAGS) \
+             $$($$($(1)_type)_$(release_type)_ARFLAGS)
+$(1)_ldflags=$$($$($(1)_type)_LDFLAGS) \
+             $$($$($(1)_type)_$$(release_type)_LDFLAGS) \
+             -L$$($($(1)_type)_prefix)/lib
+$(1)_cppflags=$$($$($(1)_type)_CPPFLAGS) \
+              $$($$($(1)_type)_$$(release_type)_CPPFLAGS) \
+              -I$$($$($(1)_type)_prefix)/include
```
Delay expansion of package variables.

tobtoht commented on .github/workflows/guix.yml:
```diff
@@ -0,0 +1,63 @@
+name: ci/gh-actions/guix
+
+on: [push, pull_request]
```
I removed guix and depends package caching from the CI workflow because it uses too much space and causes evictions for other caches (10 GB limit). We only need to run it when there is a change to `contrib/{depends,guix}`, which happens infrequently and doesn't need to complete quickly.

`on: [push, pull_request]` is temporary for testing.

tobtoht commented on .github/workflows/guix.yml:
```diff
@@ -0,0 +1,63 @@
+          path: contrib/depends/sources
+          key: sources-${{ hashFiles('contrib/depends/packages/*') }}
+      - name: install dependencies
+        run: sudo apt update; sudo apt -y install guix git ca-certificates
```
It may be worth investigating if using the installer script is faster.

https://guix.gnu.org/manual/en/html_node/Installation.html

tobtoht commented on contrib/guix/manifest.scm:
```diff
@@ -0,0 +1,271 @@
+                 (("-rpath=") "-rpath-link="))
+               #t))))))))
+
+(define-public glibc-2.27
```
The minimum glibc version for all Linux targets is now 2.27 (Ubuntu 18.04, Debian 10).

For more context, see: https://github.com/monero-project/monero/pull/9171#issuecomment-1952926105

tobtoht commented on contrib/guix/libexec/build.sh:
```diff
@@ -0,0 +1,380 @@
+      -DCMAKE_EXE_LINKER_FLAGS="${HOST_LDFLAGS}" \
+      -DCMAKE_SHARED_LINKER_FLAGS="${HOST_LDFLAGS}" \
+      -DCMAKE_SKIP_RPATH=ON \
+      -DMANUAL_SUBMODULES=1
```
We can't actually check if submodules are checked out here because we're building in an extracted source archive.

The `guix-build` script runs `git submodule update --init --recursive --progress` before starting a build to make sure submodules are checked out.

tobtoht commented:
```
423dcee4a23f697669b74188588eb1fa023b0921de1ba8c31860c6c670d6cdc2  monero-aarch64-apple-darwin-9ea73c0ff089.tar.bz2
9c40bc1ff4267f25bc2a36649bd5fd06fed2ed0fe1aadfdfca2265248ca78106  monero-aarch64-linux-android-9ea73c0ff089.tar.bz2
d7c1099361828707d2a57e5fb5407f32323bc23c5009942b0c551f17b5fd7f5c  monero-aarch64-linux-gnu-9ea73c0ff089.tar.bz2
486e6282b26dc18c4dbe1f71821bf14bd87352b6eccf29b9753f17d92bad103a  monero-arm-linux-androideabi-9ea73c0ff089.tar.bz2
85c57bde9b139303655c6f57b1305be37e613f51e0f264361b4f6e4136c42dc5  monero-arm-linux-gnueabihf-9ea73c0ff089.tar.bz2
d24844c3396c2f4b3cd0ee0d74a02d26070480361e4fdcf7b3792ac77cbb63dd  monero-i686-linux-gnu-9ea73c0ff089.tar.bz2
14265f23ae8937d233861c3371be32a1a1a6e112c5b5e193ec24f819802f5dc4  monero-i686-w64-mingw32-9ea73c0ff089.zip
49133fe551ca06965b0ace3283fd87a1d1bd4863a5c97e29bf02f892e96feeaa  monero-riscv64-linux-gnu-9ea73c0ff089.tar.bz2
75e452f349dc0996792f3551ff452301c7107080628111f5fd2364fee69e04c6  monero-source-9ea73c0ff089.tar.gz
93c0bd491b284e920cdc042e750fad2b215896f0bcff8736b921769fe75d622f  monero-x86_64-apple-darwin-9ea73c0ff089.tar.bz2
fb9a3725c547d4313577b7937ea1c34a3c4db02eae30159dd1b57e4c447f0a8c  monero-x86_64-linux-gnu-9ea73c0ff089.tar.bz2
e5aca1771d83612844e7e85aeef1d9b7897f0575757cb207e6151ab092073f25  monero-x86_64-unknown-freebsd-9ea73c0ff089.tar.bz2
58d7586c7698e64f20c584d61a98bb4140239c4b6e2184a946a93508355f3650  monero-x86_64-w64-mingw32-9ea73c0ff089.zip
```

plowsof commented:
hashes
```
423dcee4a23f697669b74188588eb1fa023b0921de1ba8c31860c6c670d6cdc2  monero-aarch64-apple-darwin-9ea73c0ff089.tar.bz2
9c40bc1ff4267f25bc2a36649bd5fd06fed2ed0fe1aadfdfca2265248ca78106  monero-aarch64-linux-android-9ea73c0ff089.tar.bz2
d7c1099361828707d2a57e5fb5407f32323bc23c5009942b0c551f17b5fd7f5c  monero-aarch64-linux-gnu-9ea73c0ff089.tar.bz2
486e6282b26dc18c4dbe1f71821bf14bd87352b6eccf29b9753f17d92bad103a  monero-arm-linux-androideabi-9ea73c0ff089.tar.bz2
85c57bde9b139303655c6f57b1305be37e613f51e0f264361b4f6e4136c42dc5  monero-arm-linux-gnueabihf-9ea73c0ff089.tar.bz2
d24844c3396c2f4b3cd0ee0d74a02d26070480361e4fdcf7b3792ac77cbb63dd  monero-i686-linux-gnu-9ea73c0ff089.tar.bz2
14265f23ae8937d233861c3371be32a1a1a6e112c5b5e193ec24f819802f5dc4  monero-i686-w64-mingw32-9ea73c0ff089.zip
49133fe551ca06965b0ace3283fd87a1d1bd4863a5c97e29bf02f892e96feeaa  monero-riscv64-linux-gnu-9ea73c0ff089.tar.bz2
75e452f349dc0996792f3551ff452301c7107080628111f5fd2364fee69e04c6  monero-source-9ea73c0ff089.tar.gz
93c0bd491b284e920cdc042e750fad2b215896f0bcff8736b921769fe75d622f  monero-x86_64-apple-darwin-9ea73c0ff089.tar.bz2
fb9a3725c547d4313577b7937ea1c34a3c4db02eae30159dd1b57e4c447f0a8c  monero-x86_64-linux-gnu-9ea73c0ff089.tar.bz2
e5aca1771d83612844e7e85aeef1d9b7897f0575757cb207e6151ab092073f25  monero-x86_64-unknown-freebsd-9ea73c0ff089.tar.bz2
58d7586c7698e64f20c584d61a98bb4140239c4b6e2184a946a93508355f3650  monero-x86_64-w64-mingw32-9ea73c0ff089.zip
```

kevcrumb commented:
I recently designed [a process for gitian-building entirely in RAM](https://monero.town/post/2485839) with only 12GB total space required, significantly less than:

> - 16GB of free disk space on the partition that /gnu/store will reside in
> - 8GB of free disk space **per platform triple**

Could Guix build results be made to match those generated by gitian?

tobtoht commented:
@kevcrumb The figures in the readme are a bit off, I'll update them. 

| Description                           | location                          |   All targets | Max. per target |
|----------------------------|-----------------------------------|--------:|----------------:|
| guix cache                 | `/gnu/store`                      |  8.0 GB |               - |
| git repository             |                                   |  0.8 GB |               - | 
| `depends` sources cache              | `SOURCES_PATH` or `contrib/depends/sources`                    |  0.9 GB |               - |
| `depends` package cache      | `BASE_CACHE` or `contrib/depends/built`                      |  1.2 GB |          290 MB |
| extracted `depends` packages | `contrib/depends/<target>`        |  6.7 GB |         1100 MB |
| build directories                 | `guix/guix-build-<commit>/build`  |  9.6 GB |          980 MB |
| outputs                    | `guix/guix-build-<commit>/output` |  1.0 GB |          100 MB |
| total                      |                                   | 28.2 GB |        12.2 GB |

If target specific build/cache directories are removed after each successfully built target, about 12.2 GB of storage is required at any time, not counting a base VM image with `guix`, `git`, `make` and `curl` installed. I could add an option that does that for storage constrained environments.

kevcrumb commented:
> about 13.1 GB of storage is required at any time

On a 16 GB system that would leave a mere 2.9 of RAM for the host's OS and the entire build process combined.

Anyway, my question was rather if we could get matching checksums between gitian and Guix, so that both methods could be employed interchangeably.

(dropping the term "reproducible builds" here and a mention of #1854, as I always struggle to find this issue)

tobtoht commented:
>Anyway, my question was rather if we could get matching checksums between gitian and Guix

No, this is not possible. We can't recreate the Gitian build environment in Guix. Even if we could, it would be unmaintainable and too limiting.

0xFFFC0000 commented on .github/workflows/guix.yml:
```diff
@@ -0,0 +1,86 @@
+          fetch-depth: 0
+          submodules: recursive
+      - name: remove bundled packages
+        run: sudo rm -rf /usr/local
```
Quick question. Do we need this if we are building with GUIX? Since the premise of GUIX is sandboxed builds.

tobtoht replied:
It's just to free up disk space. See https://github.com/monero-project/monero-gui/pull/4223

0xFFFC0000 commented on contrib/depends/patches/boost/fix_coalesce.patch:
```diff
@@ -0,0 +1,17 @@
+-    {
+-        flags darwin.compile.c++ OPTIONS $(condition) : -fcoalesce-templates ;
+-    }
++    #if $(real-version) < "4.0.0"
++    #{
++    #    flags darwin.compile.c++ OPTIONS $(condition) : -fcoalesce-templates ;
++    #}
```
Question. AFAIK we are already in process of moving away from boost 1.64. Do we need to handle this special case? Once the move to newer boost is finished, can we get rid of this part?

tobtoht replied:
This patch will be removed in #9162

iamamyth replied:
Can you inline the above as a comment, just to keep track of the upgrade path here? For example (just copied your comments and added boost bugfix version):
```
Boost 1.64.0 doesn't recognize that we're building with Clang and passes a flags that results in an error. We don't support GCC < 4.0 at all, so commenting out the lines here is fine. Patch can be dropped when we update to Boost <insert bugfix version>
```

0xFFFC0000 commented on contrib/guix/libexec/prelude.bash:
```diff
@@ -0,0 +1,84 @@
+# across time.
+time-machine() {
+    # shellcheck disable=SC2086
+    guix time-machine --url=https://github.com/tobtoht/guix.git \
```
I have nothing against using `https://github.com/tobtoht/guix.git`. But quick question, there is no way to use the official repo? Or at least report to them the failure?

tobtoht replied:
The entire repository needs to be pulled for each build. I haven't yet found a way to cache this. Pulling from Github repos is much faster (and reliable) than external repos, even if GNU fixes their rate limiting.

I can change this so only GitHub CI uses my fork and the official URL is used for local builds. Another option would be to create a fork under `monero-project/guix`.

0xFFFC0000 replied:
My vote is `monero-project/guix` .

nahuhh replied:
https://github.com/monero-project/guix

tobtoht replied:
Thank you. Can you also push the `keyring` branch? Guix needs it to authenticate commits.

nahuhh replied:
Done

tobtoht replied:
Updated.

nahuhh commented on .github/workflows/guix.yml:
```diff
@@ -0,0 +1,92 @@
+
+jobs:
+  cache-sources:
+    runs-on: ubuntu-24.04
```
Should we switch to debian (12)?

tobtoht replied:
Oh, that might fix the AppArmor issue. Good suggestion, will try.

nahuhh commented on contrib/guix/patches/glibc-2.27-riscv64-Use-__has_include-to-include-asm-syscalls.h.patch:
```diff
@@ -0,0 +1,76 @@
+Note that this has been modified from the original commit, to use __has_include
```
Is this the same issue with buggy gcc version 10-11 compiler? (Is this solved on gcc <=9 and >=12+?)

tobtoht replied:
Unrelated, patch fixes a missing include in glibc. Monero codebase doesn't contain ` __has_include__`. Log for reference:

```
../sysdeps/unix/sysv/linux/riscv/flush-icache.c:24:10: fatal error: asm/syscalls.h: No such file or directory
   24 | #include <asm/syscalls.h>
      |          ^~~~~~~~~~~~~~~~
compilation terminated.
```

tobtoht commented on contrib/guix/patches/glibc-2.27-guix-prefix.patch:
```diff
@@ -0,0 +1,22 @@
+Without ffile-prefix-map, the debug symbols will contain paths for the
```
It now does for linux, windows, and macos targets. FreeBSD and Android targets use pre-compiled x86_64 toolchains and libraries.

tobtoht commented:
```
$ uname --machine && find guix/guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
x86_64
dcf87fdab805f5a0bb77e9656024caed41330c30c3ccc35efd02fbfd3a02dd3a  guix/guix-build-ee3782d83251/output/aarch64-apple-darwin/monero-aarch64-apple-darwin-ee3782d83251.tar.bz2
150cfc02d909ef604eec0a9494538ff8344cf430dec9308a0ab785d028200832  guix/guix-build-ee3782d83251/output/aarch64-linux-gnu/monero-aarch64-linux-gnu-ee3782d83251.tar.bz2
63130e0275501be716181d2c749f8d41d51a9986389243a0a0ab5391c9755c9b  guix/guix-build-ee3782d83251/output/arm-linux-gnueabihf/monero-arm-linux-gnueabihf-ee3782d83251.tar.bz2
e53cdca1094e7059dc2603fcc157b7b9ed9dd753c623259a08e3f80282a1a6a5  guix/guix-build-ee3782d83251/output/dist-archive/monero-source-ee3782d83251.tar.gz
fc58e7c0655d6af309fbdd0e7e3c1507a370767a41428f1b40b99d53ac9b32be  guix/guix-build-ee3782d83251/output/i686-linux-gnu/monero-i686-linux-gnu-ee3782d83251.tar.bz2
12433286439b89b4fa9f4da3565b3ed5394be019653b689c57f9679fe6e904b5  guix/guix-build-ee3782d83251/output/i686-w64-mingw32/monero-i686-w64-mingw32-ee3782d83251.zip
aaeeba15997fbe2e972376e3e8cf3ac171296805d46f84c64e0b92dbd008def5  guix/guix-build-ee3782d83251/output/riscv64-linux-gnu/monero-riscv64-linux-gnu-ee3782d83251.tar.bz2
2b88cd8204c49ebb84e5612ad45c16a7f9d5ddd4acee8323ad27a04b165481d6  guix/guix-build-ee3782d83251/output/x86_64-apple-darwin/monero-x86_64-apple-darwin-ee3782d83251.tar.bz2
d619cf45fbf40b897a2b335ecba4010f3435548ce379ee71bc864024b647264a  guix/guix-build-ee3782d83251/output/x86_64-linux-gnu/monero-x86_64-linux-gnu-ee3782d83251.tar.bz2
04d2fca7317c6b3ba8e3781b3ca21c4adf12f12f31d1d51e1930bb4567622ae1  guix/guix-build-ee3782d83251/output/x86_64-w64-mingw32/monero-x86_64-w64-mingw32-ee3782d83251.zip

aarch64
dcf87fdab805f5a0bb77e9656024caed41330c30c3ccc35efd02fbfd3a02dd3a  guix/guix-build-ee3782d83251/output/aarch64-apple-darwin/monero-aarch64-apple-darwin-ee3782d83251.tar.bz2
150cfc02d909ef604eec0a9494538ff8344cf430dec9308a0ab785d028200832  guix/guix-build-ee3782d83251/output/aarch64-linux-gnu/monero-aarch64-linux-gnu-ee3782d83251.tar.bz2
63130e0275501be716181d2c749f8d41d51a9986389243a0a0ab5391c9755c9b  guix/guix-build-ee3782d83251/output/arm-linux-gnueabihf/monero-arm-linux-gnueabihf-ee3782d83251.tar.bz2
e53cdca1094e7059dc2603fcc157b7b9ed9dd753c623259a08e3f80282a1a6a5  guix/guix-build-ee3782d83251/output/dist-archive/monero-source-ee3782d83251.tar.gz
fc58e7c0655d6af309fbdd0e7e3c1507a370767a41428f1b40b99d53ac9b32be  guix/guix-build-ee3782d83251/output/i686-linux-gnu/monero-i686-linux-gnu-ee3782d83251.tar.bz2
12433286439b89b4fa9f4da3565b3ed5394be019653b689c57f9679fe6e904b5  guix/guix-build-ee3782d83251/output/i686-w64-mingw32/monero-i686-w64-mingw32-ee3782d83251.zip
aaeeba15997fbe2e972376e3e8cf3ac171296805d46f84c64e0b92dbd008def5  guix/guix-build-ee3782d83251/output/riscv64-linux-gnu/monero-riscv64-linux-gnu-ee3782d83251.tar.bz2
2b88cd8204c49ebb84e5612ad45c16a7f9d5ddd4acee8323ad27a04b165481d6  guix/guix-build-ee3782d83251/output/x86_64-apple-darwin/monero-x86_64-apple-darwin-ee3782d83251.tar.bz2
d619cf45fbf40b897a2b335ecba4010f3435548ce379ee71bc864024b647264a  guix/guix-build-ee3782d83251/output/x86_64-linux-gnu/monero-x86_64-linux-gnu-ee3782d83251.tar.bz2
04d2fca7317c6b3ba8e3781b3ca21c4adf12f12f31d1d51e1930bb4567622ae1  guix/guix-build-ee3782d83251/output/x86_64-w64-mingw32/monero-x86_64-w64-mingw32-ee3782d83251.zip
```

Matching hashes for Linux, Windows, and macOS targets on x86_64 and aarch64 based machines.

plowsof commented:
```
$ uname --machine && find guix/guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
x86_64
dcf87fdab805f5a0bb77e9656024caed41330c30c3ccc35efd02fbfd3a02dd3a  guix/guix-build-ee3782d83251/output/aarch64-apple-darwin/monero-aarch64-apple-darwin-ee3782d83251.tar.bz2
759bc05137dc076e846455ebbbcbbb6e419e52edb23fc53e40367fd782f2080f  guix/guix-build-ee3782d83251/output/aarch64-linux-android/monero-aarch64-linux-android-ee3782d83251.tar.bz2
150cfc02d909ef604eec0a9494538ff8344cf430dec9308a0ab785d028200832  guix/guix-build-ee3782d83251/output/aarch64-linux-gnu/monero-aarch64-linux-gnu-ee3782d83251.tar.bz2
14fb0cdd6c65bc8d91df5c2f6deccd94ab261d8a0d7825d401005ed702da0785  guix/guix-build-ee3782d83251/output/arm-linux-androideabi/monero-arm-linux-androideabi-ee3782d83251.tar.bz2
63130e0275501be716181d2c749f8d41d51a9986389243a0a0ab5391c9755c9b  guix/guix-build-ee3782d83251/output/arm-linux-gnueabihf/monero-arm-linux-gnueabihf-ee3782d83251.tar.bz2
e53cdca1094e7059dc2603fcc157b7b9ed9dd753c623259a08e3f80282a1a6a5  guix/guix-build-ee3782d83251/output/dist-archive/monero-source-ee3782d83251.tar.gz
fc58e7c0655d6af309fbdd0e7e3c1507a370767a41428f1b40b99d53ac9b32be  guix/guix-build-ee3782d83251/output/i686-linux-gnu/monero-i686-linux-gnu-ee3782d83251.tar.bz2
12433286439b89b4fa9f4da3565b3ed5394be019653b689c57f9679fe6e904b5  guix/guix-build-ee3782d83251/output/i686-w64-mingw32/monero-i686-w64-mingw32-ee3782d83251.zip
aaeeba15997fbe2e972376e3e8cf3ac171296805d46f84c64e0b92dbd008def5  guix/guix-build-ee3782d83251/output/riscv64-linux-gnu/monero-riscv64-linux-gnu-ee3782d83251.tar.bz2
2b88cd8204c49ebb84e5612ad45c16a7f9d5ddd4acee8323ad27a04b165481d6  guix/guix-build-ee3782d83251/output/x86_64-apple-darwin/monero-x86_64-apple-darwin-ee3782d83251.tar.bz2
d619cf45fbf40b897a2b335ecba4010f3435548ce379ee71bc864024b647264a  guix/guix-build-ee3782d83251/output/x86_64-linux-gnu/monero-x86_64-linux-gnu-ee3782d83251.tar.bz2
29558458e02045d43ca0d79322a09c284f101f965eb487e078361c60163e190c  guix/guix-build-ee3782d83251/output/x86_64-unknown-freebsd/monero-x86_64-unknown-freebsd-ee3782d83251.tar.bz2
04d2fca7317c6b3ba8e3781b3ca21c4adf12f12f31d1d51e1930bb4567622ae1  guix/guix-build-ee3782d83251/output/x86_64-w64-mingw32/monero-x86_64-w64-mingw32-ee3782d83251.zip

```

iamamyth commented:
Quite a few of the inline comments on github would perhaps better aid future maintainers as source code comments (if it's non-trivial enough to comment, why confine the comments to github, rather than the actual source?)

tobtoht commented on contrib/depends/patches/android_ndk/fix_env.patch:
```diff
@@ -0,0 +1,30 @@
+             #!/bin/bash
+             if [ "$1" != "-cc1" ]; then
+-                `dirname $0`/clang{version} {flags} "$@"
++                env -u C_INCLUDE_PATH -u CPLUS_INCLUDE_PATH -u OBJC_INCLUDE_PATH -u OBJCPLUS_INCLUDE_PATH -u CPATH -u LIBRARY_PATH `dirname $0`/clang{version} {flags} "$@"
```
This patch will be removed in #9456

tobtoht commented:
Guix `core-updates` got merged yesterday, which bumped `binutils` to `2.41`, which we'll need for #9440, so I have updated the time-machine. Another notable change is GCC `12.3.0` -> `12.4.0`.

I have also added comments to the source where useful, as was suggested by @iamamyth.

tobtoht commented:
```
$ uname --machine && find guix/guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
x86_64
8c3eb08d2412def637133c1fc03e3edbd030f77a79845312a05ad16a5936991b  guix/guix-build-f0800bf7ab91/output/aarch64-apple-darwin/monero-aarch64-apple-darwin-f0800bf7ab91.tar.bz2
790639688682dee168f4485a2a4c85cad93268c6502e513a8bfa77c83b9be2af  guix/guix-build-f0800bf7ab91/output/aarch64-linux-android/monero-aarch64-linux-android-f0800bf7ab91.tar.bz2
8291dff3d6ca0cac5a6896a2436745745e699ffa09d38bee398c9643127dd399  guix/guix-build-f0800bf7ab91/output/aarch64-linux-gnu/monero-aarch64-linux-gnu-f0800bf7ab91.tar.bz2
7ce9149eae71c5d39914f0f437f411667505f09d00852a8f2be1daac2098a220  guix/guix-build-f0800bf7ab91/output/arm-linux-androideabi/monero-arm-linux-androideabi-f0800bf7ab91.tar.bz2
ba2bf86383bc3b7447a2240477b17811d8e8ce46957139a8364927fec09401ab  guix/guix-build-f0800bf7ab91/output/arm-linux-gnueabihf/monero-arm-linux-gnueabihf-f0800bf7ab91.tar.bz2
7e82c5890574510ecfae041df8e171331cb850e650058f9d31c983fd1cea480a  guix/guix-build-f0800bf7ab91/output/dist-archive/monero-source-f0800bf7ab91.tar.gz
59e6ff448fdb5392c9963471fe06c014d66ffbb184c014ce36e4e0bd2adeda36  guix/guix-build-f0800bf7ab91/output/i686-linux-gnu/monero-i686-linux-gnu-f0800bf7ab91.tar.bz2
611c45fbdc93e7f98c291a91b8f9b7399e968890222a5e1c272f3b69922c0815  guix/guix-build-f0800bf7ab91/output/i686-w64-mingw32/monero-i686-w64-mingw32-f0800bf7ab91.zip
73a0146612932f15acbfe215b6d23228bf15fa6b9f18129f170f917adfb0dad2  guix/guix-build-f0800bf7ab91/output/riscv64-linux-gnu/monero-riscv64-linux-gnu-f0800bf7ab91.tar.bz2
005eca51f1c9312d73440bb49b99d61e0d47fcafdc21c85d473433a9d3e0d50c  guix/guix-build-f0800bf7ab91/output/x86_64-apple-darwin/monero-x86_64-apple-darwin-f0800bf7ab91.tar.bz2
7889b1aa4d87607aa295535c54aa3dff35c1d9af1f7134cd607c4ebafdc3c94e  guix/guix-build-f0800bf7ab91/output/x86_64-linux-gnu/monero-x86_64-linux-gnu-f0800bf7ab91.tar.bz2
be3df6507f5a1ff45134524582b0c430053cf5bb2d04849e4b313d93463ee4a4  guix/guix-build-f0800bf7ab91/output/x86_64-unknown-freebsd/monero-x86_64-unknown-freebsd-f0800bf7ab91.tar.bz2
e2d45a1f2cfff9fecfd45befeb748a1fbcf5206dad8171b966a40edca1653c36  guix/guix-build-f0800bf7ab91/output/x86_64-w64-mingw32/monero-x86_64-w64-mingw32-f0800bf7ab91.zip
```

Matches GitHub CI.

SChernykh commented:
This is what I got:
```
x86_64
8c3eb08d2412def637133c1fc03e3edbd030f77a79845312a05ad16a5936991b  guix/guix-build-f0800bf7ab91/output/aarch64-apple-darwin/monero-aarch64-apple-darwin-f0800bf7ab91.tar.bz2
790639688682dee168f4485a2a4c85cad93268c6502e513a8bfa77c83b9be2af  guix/guix-build-f0800bf7ab91/output/aarch64-linux-android/monero-aarch64-linux-android-f0800bf7ab91.tar.bz2
8291dff3d6ca0cac5a6896a2436745745e699ffa09d38bee398c9643127dd399  guix/guix-build-f0800bf7ab91/output/aarch64-linux-gnu/monero-aarch64-linux-gnu-f0800bf7ab91.tar.bz2
7ce9149eae71c5d39914f0f437f411667505f09d00852a8f2be1daac2098a220  guix/guix-build-f0800bf7ab91/output/arm-linux-androideabi/monero-arm-linux-androideabi-f0800bf7ab91.tar.bz2
ba2bf86383bc3b7447a2240477b17811d8e8ce46957139a8364927fec09401ab  guix/guix-build-f0800bf7ab91/output/arm-linux-gnueabihf/monero-arm-linux-gnueabihf-f0800bf7ab91.tar.bz2
7e82c5890574510ecfae041df8e171331cb850e650058f9d31c983fd1cea480a  guix/guix-build-f0800bf7ab91/output/dist-archive/monero-source-f0800bf7ab91.tar.gz
59e6ff448fdb5392c9963471fe06c014d66ffbb184c014ce36e4e0bd2adeda36  guix/guix-build-f0800bf7ab91/output/i686-linux-gnu/monero-i686-linux-gnu-f0800bf7ab91.tar.bz2
611c45fbdc93e7f98c291a91b8f9b7399e968890222a5e1c272f3b69922c0815  guix/guix-build-f0800bf7ab91/output/i686-w64-mingw32/monero-i686-w64-mingw32-f0800bf7ab91.zip
73a0146612932f15acbfe215b6d23228bf15fa6b9f18129f170f917adfb0dad2  guix/guix-build-f0800bf7ab91/output/riscv64-linux-gnu/monero-riscv64-linux-gnu-f0800bf7ab91.tar.bz2
005eca51f1c9312d73440bb49b99d61e0d47fcafdc21c85d473433a9d3e0d50c  guix/guix-build-f0800bf7ab91/output/x86_64-apple-darwin/monero-x86_64-apple-darwin-f0800bf7ab91.tar.bz2
7889b1aa4d87607aa295535c54aa3dff35c1d9af1f7134cd607c4ebafdc3c94e  guix/guix-build-f0800bf7ab91/output/x86_64-linux-gnu/monero-x86_64-linux-gnu-f0800bf7ab91.tar.bz2
be3df6507f5a1ff45134524582b0c430053cf5bb2d04849e4b313d93463ee4a4  guix/guix-build-f0800bf7ab91/output/x86_64-unknown-freebsd/monero-x86_64-unknown-freebsd-f0800bf7ab91.tar.bz2
e2d45a1f2cfff9fecfd45befeb748a1fbcf5206dad8171b966a40edca1653c36  guix/guix-build-f0800bf7ab91/output/x86_64-w64-mingw32/monero-x86_64-w64-mingw32-f0800bf7ab91.zip
```

DiosDelRayo commented:
```
JOBS=4 time ./contrib/guix/guix-build
```

```
....
building fonts directory...
building directory of Info manuals...
building profile with 32 packages...
guix environment: error: mount: mount "none" on "/tmp/guix-directory.Ile657": Permission denied
Command exited with non-zero status 1
1165.33user 44.35system 1:44:14elapsed 19%CPU (0avgtext+0avgdata 725688maxresident)k
526232inputs+3649664outputs (5937major+2972342minor)pagefaults 0swaps
```
Should I try to find the issue, simply run again or wait?

Do you need some additional input besides (k)ubuntu  24.04 based on debian version trixie/sid?

tobtoht commented:
@DiosDelRayo Oh, that's annoying. It's this Ubuntu bug: https://bugs.launchpad.net/ubuntu/+source/guix/+bug/2064115

Let me check if there is a reliable workaround / fix.

DiosDelRayo commented:
> @DiosDelRayo Oh, that's annoying. It's this Ubuntu bug: https://bugs.launchpad.net/ubuntu/+source/guix/+bug/2064115
> 
> Let me check if there is a reliable workaround / fix.
> 
> Edit: the solution here works: https://bugs.launchpad.net/ubuntu/+source/guix/+bug/2064115/comments/2
> 
> ```shell
> # in a root shell
> 
> $ apt install apparmor-utils
> $ cat <<EOL >> /etc/apparmor.d/guix
> abi <abi/4.0>,
> include <tunables/global>
> 
> profile guix /usr/bin/guix flags=(unconfined) {
>   userns,
>   include if exists <local/guix>
> }
> EOL
> 
> $ /etc/init.d/apparmor reload
> $ aa-enforce guix
> ```

After executing this steps a new issue pops up:
```
$ JOBS=4 time ./contrib/guix/guix-build 
Checking that we can connect to the guix-daemon...

Hint: If this hangs, you may want to try turning your guix-daemon off and on
      again.

/usr/lib/x86_64-linux-gnu/guix/guile: error while loading shared libraries: libguile-3.0.so.1: cannot open shared object file: Permission denied

ERR: Failed to connect to the guix-daemon, please ensure that one is running and
     reachable.
Command exited with non-zero status 1
0.06user 0.08system 0:00.14elapsed 104%CPU (0avgtext+0avgdata 6080maxresident)k
0inputs+72outputs (0major+12708minor)pagefaults 0swaps
```
dmesg since start to see what is going on with AppAmor:
```
[61489.959852] audit: type=1400 audit(1725372787.341:745): apparmor="DENIED" operation="file_inherit" class="net" profile="guix" pid=71487 comm="guix" family="unix" sock_type="stream" protocol=0 requested="send receive" denied="send receive" addr=none peer_addr=none
[61489.959856] audit: type=1400 audit(1725372787.341:746): apparmor="DENIED" operation="open" class="file" info="Failed name lookup - disconnected path" error=-13 profile="guix" name="apparmor/.null" pid=71487 comm="guix" requested_mask="wr" denied_mask="wr" fsuid=1000 ouid=0
[61489.959858] audit: type=1400 audit(1725372787.341:747): apparmor="DENIED" operation="file_inherit" class="net" profile="guix" pid=71487 comm="guix" family="unix" sock_type="stream" protocol=0 requested="send receive" denied="send receive" addr=none peer_addr=none
[61489.959860] audit: type=1400 audit(1725372787.341:748): apparmor="DENIED" operation="file_inherit" class="net" profile="guix" pid=71487 comm="guix" family="unix" sock_type="stream" protocol=0 requested="send receive" denied="send receive" addr=none peer_addr=none
[61489.959861] audit: type=1400 audit(1725372787.341:749): apparmor="DENIED" operation="file_inherit" class="net" profile="guix" pid=71487 comm="guix" family="unix" sock_type="stream" protocol=0 requested="send receive" denied="send receive" addr=none peer_addr=none
[61489.959985] audit: type=1400 audit(1725372787.341:750): apparmor="DENIED" operation="open" class="file" profile="guix" name="/etc/ld.so.cache" pid=71487 comm="guix" requested_mask="r" denied_mask="r" fsuid=1000 ouid=0
[61489.959988] audit: type=1400 audit(1725372787.341:751): apparmor="DENIED" operation="open" class="file" profile="guix" name="/usr/lib/x86_64-linux-gnu/libguile-3.0.so.1.6.0" pid=71487 comm="guix" requested_mask="r" denied_mask="r" fsuid=1000 ouid=0
[61489.959990] audit: type=1400 audit(1725372787.341:752): apparmor="DENIED" operation="getattr" class="file" profile="guix" name="/usr/lib/x86_64-linux-gnu/" pid=71487 comm="guix" requested_mask="r" denied_mask="r" fsuid=1000 ouid=0
[61489.959992] audit: type=1400 audit(1725372787.341:753): apparmor="DENIED" operation="open" class="file" profile="guix" name="/usr/lib/x86_64-linux-gnu/libguile-3.0.so.1.6.0" pid=71487 comm="guix" requested_mask="r" denied_mask="r" fsuid=1000 ouid=0
[61489.959993] audit: type=1400 audit(1725372787.341:754): apparmor="DENIED" operation="getattr" class="file" profile="guix" name="/usr/lib/x86_64-linux-gnu/" pid=71487 comm="guix" requested_mask="r" denied_mask="r" fsuid=1000 ouid=0
```
I think I can this switch AppAmor in audit only mode like with SELinux to only watch where it would break, need some minutes to read...

DiosDelRayo commented:
> I think I can this switch AppAmor in audit only mode like with SELinux to only watch where it would break, need some minutes to read...

After:
```
$ sudo -i
# aa-audit guix
# aa-complain guix
$ sudo service guix-daemon stop  # only to be sure
$ sudo service guix-daemon start
$ JOBS=4 time ./contrib/guix/guix-build
```
The output ends like before:
```
make: Leaving directory '/projects/monero-guix/contrib/depends'
INFO: Building f0800bf7ab91 for platform triple x86_64-linux-gnu:
      ...using commit timestamp: 1688324424
      ...running at most 4 jobs
      ...from worktree directory: '/projects/monero-guix'
          ...bind-mounted in container to: '/monero'
      ...in build directory: '/projects/monero-guix/guix/guix-build-f0800bf7ab91/build/distsrc-f0800bf7ab91-x86_64-linux-gnu'
          ...bind-mounted in container to: '/distsrc-base/build/distsrc-f0800bf7ab91-x86_64-linux-gnu'
      ...outputting in: '/projects/monero-guix/guix/guix-build-f0800bf7ab91/output/x86_64-linux-gnu'
          ...bind-mounted in container to: '/outdir-base/x86_64-linux-gnu'
guix environment: error: cannot create container: user namespaces unavailable
guix environment: error: is your kernel version < 3.10?
Command exited with non-zero status 1
13.81user 8.67system 0:21.89elapsed 102%CPU (0avgtext+0avgdata 52736maxresident)k
1918888inputs+5528outputs (769major+1526528minor)pagefaults 0swaps
```
dmesg since build:
```
[62485.631137] audit: type=1400 audit(1725373783.008:795): apparmor="ALLOWED" operation="file_inherit" class="net" profile="guix" pid=72664 comm="guix" family="unix" sock_type="stream" protocol=0 requested="send receive" denied="send receive" addr=none peer_addr=none
[62485.631142] audit: type=1400 audit(1725373783.008:796): apparmor="ALLOWED" operation="file_inherit" class="net" profile="guix" pid=72664 comm="guix" family="unix" sock_type="stream" protocol=0 requested="send receive" denied="send receive" addr=none peer_addr=none
[62485.631145] audit: type=1400 audit(1725373783.008:797): apparmor="ALLOWED" operation="file_inherit" class="net" profile="guix" pid=72664 comm="guix" family="unix" sock_type="stream" protocol=0 requested="send receive" denied="send receive" addr=none peer_addr=none
[62485.631146] audit: type=1400 audit(1725373783.008:798): apparmor="ALLOWED" operation="file_inherit" class="net" profile="guix" pid=72664 comm="guix" family="unix" sock_type="stream" protocol=0 requested="send receive" denied="send receive" addr=none peer_addr=none
[62485.631261] audit: type=1400 audit(1725373783.008:799): apparmor="ALLOWED" operation="open" class="file" profile="guix" name="/etc/ld.so.cache" pid=72664 comm="guix" requested_mask="r" denied_mask="r" fsuid=1000 ouid=0
[62485.631266] audit: type=1400 audit(1725373783.008:800): apparmor="ALLOWED" operation="getattr" class="file" profile="guix" name="/etc/ld.so.cache" pid=72664 comm="guix" requested_mask="r" denied_mask="r" fsuid=1000 ouid=0
[62485.631269] audit: type=1400 audit(1725373783.008:801): apparmor="ALLOWED" operation="getattr" class="file" profile="guix" name="/etc/ld.so.cache" pid=72664 comm="guix" requested_mask="r" denied_mask="r" fsuid=1000 ouid=0
[62485.631270] audit: type=1400 audit(1725373783.009:802): apparmor="ALLOWED" operation="open" class="file" profile="guix" name="/usr/lib/x86_64-linux-gnu/libguile-3.0.so.1.6.0" pid=72664 comm="guix" requested_mask="r" denied_mask="r" fsuid=1000 ouid=0
[62485.639762] audit: type=1400 audit(1725373783.017:803): apparmor="ALLOWED" operation="getattr" class="file" profile="guix" name="/usr/lib/x86_64-linux-gnu/libguile-3.0.so.1.6.0" pid=72664 comm="guix" requested_mask="r" denied_mask="r" fsuid=1000 ouid=0
[62485.639780] audit: type=1400 audit(1725373783.017:804): apparmor="ALLOWED" operation="getattr" class="file" profile="guix" name="/usr/lib/x86_64-linux-gnu/libguile-3.0.so.1.6.0" pid=72664 comm="guix" requested_mask="r" denied_mask="r" fsuid=1000 ouid=0
[62506.048225] kauditd_printk_skb: 731 callbacks suppressed
[62506.048243] audit: type=1400 audit(1725373803.425:1536): apparmor="ALLOWED" operation="open" class="file" profile="guix" name="/etc/ld.so.cache" pid=84120 comm="guix" requested_mask="r" denied_mask="r" fsuid=1000 ouid=0
[62506.048246] audit: type=1400 audit(1725373803.425:1537): apparmor="ALLOWED" operation="getattr" class="file" profile="guix" name="/etc/ld.so.cache" pid=84120 comm="guix" requested_mask="r" denied_mask="r" fsuid=1000 ouid=0
[62506.048248] audit: type=1400 audit(1725373803.425:1538): apparmor="ALLOWED" operation="getattr" class="file" profile="guix" name="/etc/ld.so.cache" pid=84120 comm="guix" requested_mask="r" denied_mask="r" fsuid=1000 ouid=0
[62506.048250] audit: type=1400 audit(1725373803.425:1539): apparmor="ALLOWED" operation="open" class="file" profile="guix" name="/usr/lib/x86_64-linux-gnu/libguile-3.0.so.1.6.0" pid=84120 comm="guix" requested_mask="r" denied_mask="r" fsuid=1000 ouid=0
[62506.048252] audit: type=1400 audit(1725373803.425:1540): apparmor="ALLOWED" operation="getattr" class="file" profile="guix" name="/usr/lib/x86_64-linux-gnu/libguile-3.0.so.1.6.0" pid=84120 comm="guix" requested_mask="r" denied_mask="r" fsuid=1000 ouid=0
[62506.048253] audit: type=1400 audit(1725373803.425:1541): apparmor="ALLOWED" operation="getattr" class="file" profile="guix" name="/usr/lib/x86_64-linux-gnu/libguile-3.0.so.1.6.0" pid=84120 comm="guix" requested_mask="r" denied_mask="r" fsuid=1000 ouid=0
[62506.048255] audit: type=1400 audit(1725373803.425:1542): apparmor="ALLOWED" operation="file_mmap" class="file" profile="guix" name="/usr/lib/x86_64-linux-gnu/libguile-3.0.so.1.6.0" pid=84120 comm="guix" requested_mask="rm" denied_mask="rm" fsuid=1000 ouid=0
[62506.048314] audit: type=1400 audit(1725373803.426:1543): apparmor="ALLOWED" operation="open" class="file" profile="guix" name="/usr/lib/x86_64-linux-gnu/libc.so.6" pid=84120 comm="guix" requested_mask="r" denied_mask="r" fsuid=1000 ouid=0
[62506.048319] audit: type=1400 audit(1725373803.426:1544): apparmor="ALLOWED" operation="getattr" class="file" profile="guix" name="/usr/lib/x86_64-linux-gnu/libc.so.6" pid=84120 comm="guix" requested_mask="r" denied_mask="r" fsuid=1000 ouid=0
[62506.048321] audit: type=1400 audit(1725373803.426:1545): apparmor="ALLOWED" operation="getattr" class="file" profile="guix" name="/usr/lib/x86_64-linux-gnu/libc.so.6" pid=84120 comm="guix" requested_mask="r" denied_mask="r" fsuid=1000 ouid=0
```

plowsof commented:
```
uname --machine && find guix/guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
x86_64
8c3eb08d2412def637133c1fc03e3edbd030f77a79845312a05ad16a5936991b  guix/guix-build-f0800bf7ab91/output/aarch64-apple-darwin/monero-aarch64-apple-darwin-f0800bf7ab91.tar.bz2
790639688682dee168f4485a2a4c85cad93268c6502e513a8bfa77c83b9be2af  guix/guix-build-f0800bf7ab91/output/aarch64-linux-android/monero-aarch64-linux-android-f0800bf7ab91.tar.bz2
8291dff3d6ca0cac5a6896a2436745745e699ffa09d38bee398c9643127dd399  guix/guix-build-f0800bf7ab91/output/aarch64-linux-gnu/monero-aarch64-linux-gnu-f0800bf7ab91.tar.bz2
7ce9149eae71c5d39914f0f437f411667505f09d00852a8f2be1daac2098a220  guix/guix-build-f0800bf7ab91/output/arm-linux-androideabi/monero-arm-linux-androideabi-f0800bf7ab91.tar.bz2
ba2bf86383bc3b7447a2240477b17811d8e8ce46957139a8364927fec09401ab  guix/guix-build-f0800bf7ab91/output/arm-linux-gnueabihf/monero-arm-linux-gnueabihf-f0800bf7ab91.tar.bz2
7e82c5890574510ecfae041df8e171331cb850e650058f9d31c983fd1cea480a  guix/guix-build-f0800bf7ab91/output/dist-archive/monero-source-f0800bf7ab91.tar.gz
59e6ff448fdb5392c9963471fe06c014d66ffbb184c014ce36e4e0bd2adeda36  guix/guix-build-f0800bf7ab91/output/i686-linux-gnu/monero-i686-linux-gnu-f0800bf7ab91.tar.bz2
611c45fbdc93e7f98c291a91b8f9b7399e968890222a5e1c272f3b69922c0815  guix/guix-build-f0800bf7ab91/output/i686-w64-mingw32/monero-i686-w64-mingw32-f0800bf7ab91.zip
73a0146612932f15acbfe215b6d23228bf15fa6b9f18129f170f917adfb0dad2  guix/guix-build-f0800bf7ab91/output/riscv64-linux-gnu/monero-riscv64-linux-gnu-f0800bf7ab91.tar.bz2
005eca51f1c9312d73440bb49b99d61e0d47fcafdc21c85d473433a9d3e0d50c  guix/guix-build-f0800bf7ab91/output/x86_64-apple-darwin/monero-x86_64-apple-darwin-f0800bf7ab91.tar.bz2
7889b1aa4d87607aa295535c54aa3dff35c1d9af1f7134cd607c4ebafdc3c94e  guix/guix-build-f0800bf7ab91/output/x86_64-linux-gnu/monero-x86_64-linux-gnu-f0800bf7ab91.tar.bz2
be3df6507f5a1ff45134524582b0c430053cf5bb2d04849e4b313d93463ee4a4  guix/guix-build-f0800bf7ab91/output/x86_64-unknown-freebsd/monero-x86_64-unknown-freebsd-f0800bf7ab91.tar.bz2
e2d45a1f2cfff9fecfd45befeb748a1fbcf5206dad8171b966a40edca1653c36  guix/guix-build-f0800bf7ab91/output/x86_64-w64-mingw32/monero-x86_64-w64-mingw32-f0800bf7ab91.zip

```

0xFFFC0000 commented:
FWIW:

> @DiosDelRayo Oh, that's annoying. It's this Ubuntu bug: https://bugs.launchpad.net/ubuntu/+source/guix/+bug/2064115
> 
> Let me check if there is a reliable workaround / fix.
> 
> Edit: the solution here works: https://bugs.launchpad.net/ubuntu/+source/guix/+bug/2064115/comments/2
> 
> ```shell
> # in a root shell
> 
> $ apt install apparmor-utils
> $ cat <<EOL >> /etc/apparmor.d/guix
> abi <abi/4.0>,
> include <tunables/global>
> 
> profile guix /usr/bin/guix flags=(unconfined) {
>   userns,
>   include if exists <local/guix>
> }
> EOL
> 
> $ /etc/init.d/apparmor reload
> $ aa-enforce guix
> ```

It seems there is a bug with apparmor: 

```
# >> sudo aa-enforce guix
Traceback (most recent call last):
  File "/usr/sbin/aa-enforce", line 33, in <module>
    tool.cmd_enforce()
  File "/usr/lib/python3/dist-packages/apparmor/tools.py", line 134, in cmd_enforce
    for (program, prof_filename, output_name) in self.get_next_for_modechange():
  File "/usr/lib/python3/dist-packages/apparmor/tools.py", line 97, in get_next_for_modechange
    aaui.UI_Info(_('Profile for %s not found, skipping') % output_name)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: 'NoneType' object is not callable


An unexpected error occurred!

For details, see /tmp/apparmor-bugreport-h96dp4_o.txt
Please consider reporting a bug at https://gitlab.com/apparmor/apparmor/-/issues
```

Appears to me has been fixed here:  https://gitlab.com/apparmor/apparmor/-/issues/411

0xFFFC0000 reviewed (commented):
`i686-linux-gnu` Result:

**Compiled** with `guix` for `HOST=i686-linux-gnu`. 
Unit tests ran **without** issue on `i686-linux-gnu`.

0xFFFC0000 commented:
`aarch64-linux-gnu` Result:

**Compiled** with `guix` for `HOST=aarch64-linux-gnu`. 
Unit tests ran **without** issue on `aarch64-linux-gnu`.

0xFFFC0000 reviewed (commented):
`aarch64-linux-gnu` Result:

**Compiled** with `guix` for `HOST=aarch64-linux-gnu`. 
Unit tests ran **without** issue on `aarch64-linux-gnu`.

0xFFFC0000 reviewed (commented):
`x86_64-unknown-freebsd` Result:

**Compiled** with `guix` for `HOST=x86_64-unknown-freebsd`. 
Unit tests ran **without** issue on `x86_64-unknown-freebsd`.

0xFFFC0000 reviewed (commented):
`aarch64-apple-darwin` Result:

**Compiled** with `guix` for `HOST=aarch64-apple-darwin `. 
Unit tests ran **without** issue on `aarch64-apple-darwin`.

0xFFFC0000 reviewed (approved):
Thanks for working on this. 


`build-macos` [1] and `build-windows` [2] failure are unrelated and has been addressed in master. 


1. https://github.com/monero-project/monero/actions/runs/10662553939/job/29549974869?pr=8929
2. https://github.com/monero-project/monero/actions/runs/10662553939/job/29549975090?pr=8929

tobtoht commented:
@iamamyth Are you perchance capable and willing to give this PR a full review when time permits?

This PR is the foundation for #9440, which will allow us to ship FCMP++, and about a dozen other PRs. It is critical that build system changes are merged ahead of time to avoid unnecessary delays in the hardfork timeline.

iamamyth commented:
@tobtoht I think I'll have time to review this within the next week (i.e. 7 days). I've already been fiddling with the build system on a local branch (have code ready to automate make job count, but don't want to PR it because there's a huge backlog as-is) and inspecting other build system PRs, so I should have enough context.

iamamyth commented on .github/workflows/guix.yml:
```diff
@@ -0,0 +1,92 @@
+      - name: install dependencies
+        run: sudo apt update; sudo apt -y install guix git ca-certificates apparmor-utils
+      - name: fix apparmor
+        run: sudo cp .github/workflows/guix /etc/apparmor.d/guix; sudo /etc/init.d/apparmor reload; sudo aa-enforce guix || true
```
It would make more sense to me if you just inlined this file because nothing else needs it. If you want to keep this as a single step, you can just put a trivial three-command script here, which will be run via the shell (e.g. tee with a heredoc and then the reload/enforce).

iamamyth commented on contrib/guix/libexec/prelude.bash:
```diff
@@ -0,0 +1,104 @@
+
+same_dir() {
+    local resolved1 resolved2
+    resolved1="$(bash_realpath "${1}")"
```
Does `realpath -e` not work on these machines or something?

tobtoht replied:
I don't we actually need a portable implementation here, since `realpath` is available by default on practically every Linux distribution.

iamamyth commented on contrib/guix/guix-attest:
```diff
@@ -0,0 +1,263 @@
+echo "Looking for build output SHA256SUMS fragments in ${LOGDIR_BASE}"
+
+shopt -s nullglob
+sha256sum_fragments=( "$LOGDIR_BASE"/*/SHA256SUMS.part ) # This expands to an array of directories...
```
What about just using `find` here?

iamamyth replied:
Sorry, didn't get a chance to finish my thoughts on this file earlier: I think you can structure the pipeline a bit differently by doing separate traversals for each file category (adds a dir scan to hopefully simplify the logic). Then, you can use the `find -print0 | xargs -0` pattern, and your error handling becomes a check on whether the outputs are empty; partial example:
```
find "$LOGDIR_BASE" -name "SHA256SUMS.part" -print0 | xargs -0 sort -u /dev/null | basenameify_SHA256SUMS | sort -k2
```

Because nocodesigned fragments are mandatory, you end up with a pretty nice overall program flow:
1. Process/validate nocodesigned fragments (error if empty file).
2. Process/validate codesigned fragments (info if empty file).

tobtoht replied:
I simplified the script based on your feedback. I also removed the {non,}codesigned logic, since I don't think we have plans to add codesigning for release binaries in the near future. Should that change, we can add the logic then.

tobtoht replied:
Forgot to adjust `guix-verify` based on these changes, will do that later.

iamamyth commented:
Just an FYI, I'm ~70% done reviewing this PR. I'll post a summary comment when I complete my first pass review, but, so, far, everything looks good.

iamamyth commented on contrib/depends/Makefile:
```diff
@@ -111,7 +116,7 @@ $(host_arch)_$(host_os)_native_toolchain?=$($(host_os)_native_toolchain)
 
 toolchain_path=$($($(host_arch)_$(host_os)_native_toolchain)_prefixbin)
-final_build_id_long+=$(shell $(build_SHA256SUM) toolchain.cmake.in)
+final_build_id_long+=:[sha256sum]:$(shell $(build_SHA256SUM) toolchain.cmake.in)
```
Is there a particular convention you want these templates to follow? Seems it's a bit all over the map, as I see `:[key]:value` here, `:[key]value:` in contrib/depends/funcs.mk, etc. I suggest picking a convention which both a human and machine can read easily (i.e. a simple procedure, which can easily be implemented in code and by the human brain, not necessarily the fastest). With that in mind, if you want to use `:` as a separator between pairs, it probably shouldn't also be used after the key.

tobtoht replied:
The colon separator is replaced with a newline in the final output (`depends-hashes.txt`) and is easily diff-able:

```
[boost_all_file_checksums]
5419a1f3ae6346deeb4ad5d60964f09a2afd443e75f86bff52032cfd6b1e5189 
2948dc359030cbd8fe411bc544212d7cf7c3308643aec3b550ddda17d512802f 
884976d97b7a386c4c0c4aeb3e0ffe40848481d0fd5f379044815ca1e24f5622 
0ddb4c9faeb4443251af83ef8924cc2579c225d8e3cc7a1b65fc206e22b0b4f2 
20a3fff6d950af9b70daf3ab005f6ba19c0b069970470e1ca81ee1e626ffe680 
897adbccc9b4bd22585dde1c443a119eaa5e01641b43b3d5960f7137197b3b54 
7284e8b6e1471f20c0e12d3f99f4301b1bc4368fd5b7d330557386939a136d24

[recipe]
unbound-1.19.1-8a76263f0ba3f0bc927efe0b29ff3067a58dbcf4d45126b42cef26bffde55c26-Release
[deps]
expat-2.6.0-ea84dd3e3694465ec4a470c96e79089ed11cbc13f43a3201094dda24c0c11814 
openssl-3.0.13-348c61a404a4575b3270d742dce1e58b5c6b0e140d67030fe20b5c15d5e6cd38
[x86_64_linux_id]
/gnu/store/03rfvpz53ab1klw20j5padpxmygcmksl-profile

[sha256sum]
1e82b195b37594b1753317b0b864dd036af3e1b7812dc6825df7362872158933  toolchain.cmake.in
```

I can probably turn this into valid yaml with minor modifications to the code, so that it's easily machine readable.

iamamyth replied:
No need to make it valid YAML, the format you have is easily parseable, I misread one of the foreach statements (didn't notice it was adding a newline with echo).

iamamyth commented on contrib/guix/libexec/build.sh:
```diff
@@ -0,0 +1,410 @@
+if [ ! -e "$GIT_ARCHIVE" ]; then
+    mkdir -p "$(dirname "$GIT_ARCHIVE")"
+    git ls-files --recurse-submodules \
+    | cat \
```
Why would `cat` be helpful/necessary here?

tobtoht replied:
It isn't, oops.

iamamyth commented on contrib/guix/libexec/build.sh:
```diff
@@ -0,0 +1,410 @@
+###########################
+
+# CFLAGS
+HOST_CFLAGS="-O2"
```
I wonder if this would be clearer if you just had a `*` in the case statement and pushed these two lines into the case, where appropriate. Is the `find` command even helpful on non-linux/gnu hosts?

tobtoht replied:
I cleaned this up. We have logic to set the optimization level for Release builds inside `CMakeLists.txt`, so setting it here is redundant/unnecessary.

iamamyth commented on contrib/guix/libexec/build.sh:
```diff
@@ -0,0 +1,410 @@
+# CXXFLAGS
+HOST_CXXFLAGS="$HOST_CFLAGS"
+
+case "$HOST" in
```
This seems more like an `if` than a `case`.

tobtoht replied:
I'd like to keep it structured like this so it's easy to add new target specific cxxflags in the future.

iamamyth commented on contrib/guix/guix-build:
```diff
@@ -0,0 +1,471 @@
+# shellcheck disable=SC2046,SC2086
+{
+    # Get depends precious dir definitions from depends
+    for precious_dir_name in $depends_precious_dir_names; do
```
You could make this a single loop over all the `precious_dir_names` with an if/else based on containment in `depends_precious_dir_names`, such that the action is shared (echo), and only the resolution mechanism differs.

iamamyth commented on contrib/guix/guix-build:
```diff
@@ -0,0 +1,471 @@
+#
+#   HOST: The current platform triple we're building for
+#
+outdir_for_host() {
```
Nothing seems to use the `SUFFIX` arg.

iamamyth commented on contrib/guix/guix-build:
```diff
@@ -0,0 +1,471 @@
+#
+#   HOST: The current platform triple we're building for
+#
+profiledir_for_host() {
```
Here as well, SUFFIX support could be removed.

iamamyth commented on contrib/guix/guix-clean:
```diff
@@ -0,0 +1,83 @@
+#
+under_dir() {
+    local path_residue
+    path_residue="${2##"${1}"}"
```
This function is easy to get wrong and could use a test suite: For safety, you'd want to check that the child starts with "$parent/". Otherwise, as it stands, you get results such as:
```
$ under_dir /tmp /tmpx/config.py  
x/config.py
```

Also, while this function name is clear enough in the local context, if you end up making it a library function, I'd consider something such as `path_contains`, to emphasize that it's just testing path-based containment, not actual containment. As for testing, you could just put this function in its own file, then source it here and via the test (dumping it into one huge library has the disadvantage that, as the library grows, the set of false implicit dependencies grows, which makes scripts more bloated).

iamamyth replied:
One other thing: Normally it's helpful to, in addition to adding a "/" to parent, support the case where parent already contains one (pretty easy to do by e.g. dropping all trailing slashes and then adding one).

tobtoht replied:
I fixed the edge cases in this function. Since the function is only used here, I didn't want to move it.

I believe I have addresses all your comments.

iamamyth commented on contrib/guix/libexec/prelude.bash:
```diff
@@ -0,0 +1,101 @@
+    done
+}
+
+check_tools cat env readlink dirname basename git
```
I'm a bit confused as to the policy here: Other files also appear to check for a superset of these commands. It would make sense to me if the prelude checked for tools required by the prelude (realpath, cat, git), and each script checked for the tools it needs.

iamamyth commented on contrib/guix/guix-verify:
```diff
@@ -0,0 +1,124 @@
+        echo "Hint: Either the signature is invalid or the public key is missing"
+        echo ""
+        failure=1
+    elif ! diff --report-identical "$compare_manifest" "$current_manifest" 1>&2; then
```
Might want to explicitly add `--text` here.

iamamyth commented:
Full review complete. The only concerning issue I saw was the `under_dir` logic.

iamamyth commented on contrib/guix/guix-attest:
```diff
@@ -0,0 +1,197 @@
+################
+
+cmd_usage() {
+cat <<EOF
```
Just going to post this here as a line comment, but it applies throughout: Some folks have very strong opinions on using cat rather than echo/printf in these situations; I don't really care, but I will just note that
```
cat <<EOF
...
EOF
```

can be expressed as:

```
echo -n "\
...
"
```

where `...` is a placeholder for a multi-line string.

hinto-janai commented:
```
x86_64
ba1f8586981975dcddc33adabe07d0ebc1e7aebbfa9ba21ba3c0ac2f6d49acff  guix/guix-build-b2ff7641a538/output/aarch64-apple-darwin/monero-aarch64-apple-darwin-b2ff7641a538.tar.bz2
039473552e9aeffc36e6073d062200393229d76780973bccdaa20f584a58b9a2  guix/guix-build-b2ff7641a538/output/aarch64-linux-android/monero-aarch64-linux-android-b2ff7641a538.tar.bz2
7cffeeb1eccb4ec1849d87aefa2865f16c908090f13132af665dd361397d5ca6  guix/guix-build-b2ff7641a538/output/aarch64-linux-gnu/monero-aarch64-linux-gnu-b2ff7641a538.tar.bz2
993e70ce72216c192efcf091f6b4f8652c07fbaaeae0bcd0a1f3a00c808f021d  guix/guix-build-b2ff7641a538/output/arm-linux-androideabi/monero-arm-linux-androideabi-b2ff7641a538.tar.bz2
e32427d93203220569c3a5938c0997a3044a2cec186b7c4282778bf60576b5f2  guix/guix-build-b2ff7641a538/output/arm-linux-gnueabihf/monero-arm-linux-gnueabihf-b2ff7641a538.tar.bz2
468457fbc9df6d8ccc18114dde9769fed8f8ecea272c8710b299afc3ad93f776  guix/guix-build-b2ff7641a538/output/dist-archive/monero-source-b2ff7641a538.tar.gz
cfb3979ab62e9832a7d42a4b2316d8b5abf5361411a11364a09211d71e909149  guix/guix-build-b2ff7641a538/output/i686-linux-gnu/monero-i686-linux-gnu-b2ff7641a538.tar.bz2
9f9fbe6aea90dc940bd9b32dd26660ddf5204038d6147bc5e857c0c1758db3a2  guix/guix-build-b2ff7641a538/output/riscv64-linux-gnu/monero-riscv64-linux-gnu-b2ff7641a538.tar.bz2
212ebd1d2a635a0b81eb4d208e4b23e3fc7c157a1d7ec5cf178bc49503833e00  guix/guix-build-b2ff7641a538/output/x86_64-apple-darwin/monero-x86_64-apple-darwin-b2ff7641a538.tar.bz2
ec8f697b6e8f3640ccf43833784b97adaebeec660362e7ceac112b779ff43bd9  guix/guix-build-b2ff7641a538/output/x86_64-linux-gnu/monero-x86_64-linux-gnu-b2ff7641a538.tar.bz2
137bf925f369d8e886255e87620b45d9f6e4bc8a94706cb26a15dd7fafd7190e  guix/guix-build-b2ff7641a538/output/x86_64-unknown-freebsd/monero-x86_64-unknown-freebsd-b2ff7641a538.tar.bz2
606945c4075c32593eadf06ccea4b2ccf0a0a5fe766bb7d1bdd08389bae2323b  guix/guix-build-b2ff7641a538/output/x86_64-w64-mingw32/monero-x86_64-w64-mingw32-b2ff7641a538.zip
```

tobtoht commented:
@hinto-janai Nice, these match the hashes from the [CI](https://github.com/tobtoht/monero/actions/runs/12223635446) job. To verify download the log bundle and run:

```
find . -name "SHA256SUMS.part" -print0 | xargs -0 sort -u /dev/null | sort -k2
```

The hashes posted earlier on this PR were for a different commit. We could use more builds on #9467, if you're up for that.
